File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/main/java/com/opspilot/strategies/ReactStrategy.java
package com.opspilot.strategies;

import com.opspilot.domain.ExecutionMetrics;
import com.opspilot.domain.ReasoningStrategy;
import com.opspilot.domain.StrategyRunInput;
import com.opspilot.domain.StrategyResult;
import com.opspilot.domain.TraceEvent;
import com.opspilot.llm.ChatMessage;
import com.opspilot.llm.ChatModel;
import com.opspilot.llm.ModelResponse;
import com.opspilot.llm.ToolCall;
import com.opspilot.tools.Tool;

import java.util.ArrayList;
import java.util.List;

/**
 * Porte simplificado de agents/react.ts / strategies/react.ts: loop
 * observacao->acao unico agente, usado como estrategia de referencia/baseline
 * ao lado do TeamStrategy multiagente (o foco desta unidade).
 */
public class ReactStrategy implements ReasoningStrategy {

    public static final String SYSTEM_PROMPT =
            "Voce e o OpsPilot, copiloto de plantao de incidentes. Use as ferramentas disponiveis "
                    + "para diagnosticar e agir; responda com a resposta final quando tiver o suficiente.";

    private final ChatModel model;
    private final List<Tool> tools;
    private final int maxIterations;

    public ReactStrategy(ChatModel model, List<Tool> tools) {
        this(model, tools, 10);
    }

    public ReactStrategy(ChatModel model, List<Tool> tools, int maxIterations) {
        this.model = model;
        this.tools = tools;
        this.maxIterations = maxIterations;
    }

    @Override
    public String name() {
        return "react";
    }

    @Override
    public StrategyResult run(StrategyRunInput input) {
        long startedAt = System.currentTimeMillis();
        List<ChatMessage> messages = new ArrayList<>(List.of(
                ChatMessage.system(SYSTEM_PROMPT),
                ChatMessage.user(input.message())));
        List<TraceEvent> trace = new ArrayList<>();
        int llmCalls = 0;
        String answer = "";

        for (int iteration = 0; iteration < maxIterations; iteration++) {
            ModelResponse response = model.invoke(messages, tools);
            llmCalls++;

            if (response.toolCalls().isEmpty()) {
                answer = response.content();
                trace.add(TraceEvent.answer("react", answer));
                break;
            }

            for (ToolCall call : response.toolCalls()) {
                trace.add(TraceEvent.action("react", call.toolName() + " " + call.args()));
                String observation = executeTool(call);
                trace.add(TraceEvent.observation("react", observation));
                messages.add(new ChatMessage(ChatMessage.Role.TOOL, observation));
            }
        }

        return new StrategyResult(
                answer, trace, new ExecutionMetrics(llmCalls, System.currentTimeMillis() - startedAt));
    }

    private String executeTool(ToolCall call) {
        for (Tool tool : tools) {
            if (tool.name().equals(call.toolName())) {
                return tool.execute(call.args());
            }
        }
        return "Erro: ferramenta \"" + call.toolName() + "\" nao encontrada.";
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/strategies/react_strategy.go
// Package strategies contém a estratégia ReAct de referência (porte
// simplificado de agents/react.ts / strategies/react.ts): loop
// observação->ação single-agent, usada como baseline ao lado da estratégia
// multiagente TeamStrategy (pacote team), que é o foco desta unidade.
package strategies

import (
	"fmt"
	"time"

	"opspilot/domain"
	"opspilot/llm"
	"opspilot/tools"
)

const ReactSystemPrompt = "Voce e o OpsPilot, copiloto de plantao de incidentes. Use as ferramentas disponiveis " +
	"para diagnosticar e agir; responda com a resposta final quando tiver o suficiente."

const defaultMaxIterations = 10

// ReactStrategy é o porte de ReactStrategy.java / react.ts.
type ReactStrategy struct {
	Model         llm.ChatModel
	Tools         []tools.Tool
	MaxIterations int
}

var _ domain.ReasoningStrategy = (*ReactStrategy)(nil)

// NewReactStrategy cria a estratégia com o teto de iterações default (10),
// igual ao construtor de dois argumentos do porte Java.
func NewReactStrategy(model llm.ChatModel, toolset []tools.Tool) *ReactStrategy {
	return NewReactStrategyWithMaxIterations(model, toolset, defaultMaxIterations)
}

func NewReactStrategyWithMaxIterations(model llm.ChatModel, toolset []tools.Tool, maxIterations int) *ReactStrategy {
	return &ReactStrategy{Model: model, Tools: toolset, MaxIterations: maxIterations}
}

func (s *ReactStrategy) Name() string { return "react" }

func (s *ReactStrategy) Run(input domain.StrategyRunInput) (domain.StrategyResult, error) {
	startedAt := time.Now()
	messages := []llm.ChatMessage{
		llm.SystemMessage(ReactSystemPrompt),
		llm.UserMessage(input.Message),
	}
	var trace []domain.TraceEvent
	llmCalls := 0
	answer := ""

	maxIterations := s.MaxIterations
	if maxIterations <= 0 {
		maxIterations = defaultMaxIterations
	}

	for iteration := 0; iteration < maxIterations; iteration++ {
		response, err := s.Model.Invoke(messages, s.Tools)
		llmCalls++
		if err != nil {
			return domain.StrategyResult{}, err
		}

		if len(response.ToolCalls) == 0 {
			answer = response.Content
			trace = append(trace, domain.Answer("react", answer))
			break
		}

		for _, call := range response.ToolCalls {
			trace = append(trace, domain.Action("react", fmt.Sprintf("%s %v", call.ToolName, call.Args)))
			observation := executeTool(s.Tools, call)
			trace = append(trace, domain.Observation("react", observation))
			messages = append(messages, llm.ToolMessage(observation))
		}
	}

	return domain.StrategyResult{
		Answer: answer,
		Trace:  trace,
		Metrics: domain.ExecutionMetrics{
			LLMCalls:  llmCalls,
			LatencyMs: time.Since(startedAt).Milliseconds(),
		},
	}, nil
}

func executeTool(toolset []tools.Tool, call llm.ToolCall) string {
	for _, tool := range toolset {
		if tool.Name() == call.ToolName {
			result, err := tool.Execute(call.Args)
			if err != nil {
				return "Erro: " + err.Error()
			}
			return result
		}
	}
	return fmt.Sprintf("Erro: ferramenta %q nao encontrada.", call.ToolName)
}
